Further enhancements to core/test-extension and extension-authors-guide

starksm64 opened the pull request:
The PR implements more of the testing outlined in #780 and adds a new Sample Test Extension section to the extension-authors-guide that illustrates several common extension tasks. A TOC was added and the header formatting was corrected to make the document more navigable.

gsmet commented on core/test-extension/runtime/pom.xml:
```diff
@@ -53,6 +53,15 @@
+        <dependency>
+            <groupId>org.glassfish.jaxb</groupId>
+            <artifactId>jaxb-runtime</artifactId>
+        </dependency>
```
I think it should be done in the `quarkus-jaxb` extension and you should depend on it.

starksm64 replied:
I would be concerned the core module is depending on too many modules outside of the core?

Sanne replied:
what's the resolution here?  There's lots of good stuff in this PR, I'd rather merge it than wait.

Is it important we have the quarkus-jaxb module if this jaxb dep is only used for testing? @gsmet could you open a separate issue if you think so? I'll merge soon if I don't see blockers :)

starksm64 replied:
Right, I would rather get the content out and then fix the dependency issue for the next release once I can verify that it does not introduce the issues like the native image build dependency caused. I created an issue for this. #1819

gsmet commented on core/test-extension/runtime/src/main/java/io/quarkus/extest/runtime/subst/KeyProxy.java:
```diff
@@ -1,5 +1,8 @@
 
+/**
+ *
+ */
```
Hmmm? Either we should have some javadoc or not.

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -1,9 +1,10 @@
 :numbered:
 :sectnums:
 :sectnumlevels: 4
+:toc:
```
We will need to check how it looks like on the website.

mkouba replied:
We should IMHO add TOC to all guides - recently, I created https://github.com/quarkusio/quarkus/issues/1490

gsmet commented on core/test-extension/deployment/src/test/java/io/quarkus/extest/ConfiguredBean.java:
```diff
@@ -1,20 +1,19 @@
-import javax.enterprise.context.ApplicationScoped;
 import javax.enterprise.event.Observes;
-import javax.inject.Inject;
+import javax.enterprise.inject.spi.CDI;
```
I don't see this import used (just checking on GH so I might be missing something)

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+author will typically need to perform using the test-extension code to illustrate how the task could be done.
+
+=== Features and Capabilities
+When you startup a {project-name} instance, you will see a line like the one highlighted in this example:
```
s/startup/start/ ?

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+     * 
+     * @return test-extension feature build item
+     */
+    @BuildStep(providesCapabilities = "io.quarkus.test-extension")
```
Not sure it helps having the capability defined here.

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+    }
+----
+
+The feature name can be used to select the extension when creating a project using the {project-name} maven plugin as shown in this example taken from the Writing JSON REST Services guide where the "resteasy-jsonb" feature is referenced:
```
AFAIK, that is not true. The feature name is a totally different beast and is only used for display.

mkouba replied:
I think so too - names and labels used by CLI are currently hard-coded in https://github.com/quarkusio/quarkus/blob/master/devtools/common/src/main/filtered/extensions.json

gsmet replied:
@starksm64 could you address this one? The statement is false AFAIK. The feature is just there to be displayed in the feature list at startup.

starksm64 replied:
Ok, right. I have corrected that section to indicate that the feature name should match a label in the extensions.json entry so that there is a correspondence between the displayed name and the label which is actually what is used to select the extension.

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+...
+----
+<1> Register the annotation class and CDI default scope using the Jandex DotName class.
+<2> ConfiguredBean will be processed by the CDI layer the same as a bean annotated with the CDI standard @ApplicationScoped.
```
Some backticks to enclose the classnames would be nice.

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+<2> ConfiguredBean will be processed by the CDI layer the same as a bean annotated with the CDI standard @ApplicationScoped.
+
+=== Parsing Config to Objects
+One of the main things an extension is likely to do is completely separate the configuration phase of behavior from the runtime phase. Frameworks often do parsing/load of configuration on startup that can be done during build time to both reduce the runtime dependencies on frameworks like xml parsers as well as reducing the startup time the parsing incurs.
```
So I'm not a native speaker so it's worth what it's worth but I find these 2 sentences hard to parse, especially the second one. And I'm not sure I see what Quarkus brings here.

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+
+----
+<1> Look for a config.xml classpath resource
+<2> If found, parse using JAXB context for `XmlConfig.class`
```
So while having an example of an XML file is nice, I think a lot of people use YAML nowadays.

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+=== Scanning Deployments Using Jandex
+If your extension defines annotations or interfaces that mark beans needing to be processed, you can locate these beans using the Jandex API, a Java annotation indexer and offline reflection library. The following `TestProcessor#scanForBeans` method showns how to find the beans annotated with our `@TestAnnotation` tha also implement the `IConfigConsumer` interface:
+
+.x
```
Looks like you're missing a title.

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+====
+
+=== Scanning Deployments Using Jandex
+If your extension defines annotations or interfaces that mark beans needing to be processed, you can locate these beans using the Jandex API, a Java annotation indexer and offline reflection library. The following `TestProcessor#scanForBeans` method showns how to find the beans annotated with our `@TestAnnotation` tha also implement the `IConfigConsumer` interface:
```
s/showns/shows/
s/tha also/that also/

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+<1> Depend on a `BeanArchiveIndexBuildItem` to have the build step be run after the deployment has been indexed.
+<2> Retrieve the index.
+<3> Find all beans annotated with `@TestAnnotation`.
+<4> Determine which of these beans also has the `IConfigConsumer` interface.
```
s/also has/also have/ ?

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+<5> Save the bean class in a `TestBeanBuildItem` for use in a latter RUNTIME_INIT build step that will interact with the bean instances.
+
+=== Interacting With Extension Beans
+You can use the `io.quarkus.arc.runtime.BeanContainer` interface to interace with your extension beans. The following `configureBeans` methods illustrate interacting with the beans scanned for in the previous section:
```
s/interace/interact/

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+        log.infof("configureBeans, instance=%s\n", instance);
+    }
+----
+<1> Consume the `TestBeanBuildItem`s produced from the scanning build step.
```
I wonder if you need double backticks here as you have a `s` just after.

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+}
+----
+<1> Call into the runtime template to record the creation of the service.
+<2> Using the parsed `XmlConfig` instance, create a instance of `RuntimeXmlConfigService` and wrap it in a `RuntimeValue`. Use a `RuntimeValue` wrapper for non-interface objects that are non-proxiable.
```
s/a instance/an instance/

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+<1> We consume a ShutdownContextBuildItem to register the service shutdown.
+<2> We consume the previously initialized service captured in `RuntimeServiceBuildItem`.
+<3> Call the runtime template to record the service start invocation.
+<4> Runtime template retrieves the service instance reference and calls it's `startService` method.
```
s/it's/its/

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+The code for the `RuntimeXmlConfigService` can be viewed here:
+{quarkus-blob-url}/core/test-extension/runtime/src/main/java/io/quarkus/extest/runtime/RuntimeXmlConfigService.java[RuntimeXmlConfigService.java]
+
+The testcase for validating that the `RuntimeXmlConfigService` has started can be found in the `testRuntimeXmlConfigService` test of `ConfiguredBeanTest` and `NativeImageIT`.
```
s/testcase/test case/ maybe?

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+The testcase for validating that the `RuntimeXmlConfigService` has started can be found in the `testRuntimeXmlConfigService` test of `ConfiguredBeanTest` and `NativeImageIT`.
+
+=== Register Resources for Use in Native Image
+Not all configuration or resources can be consumed at build time. If you have classpath resources that the runtime needs to access, you need to inform the build phase that these resource need to be copied into the native image. This is done by producing one or more `SubstrateResourceBuildItem` or `SubstrateResourceBundleBuildItem` in the case of resource bundles. Examples of this are shown in thi ...
```
s/these resource/these resources/

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+<3> Indicate that the "javax.xml.bind.Messages" resource bundle should be copied into native image.
+
+=== Object Substitution
+Objects created during the build phase that are passed into the runtime need to have a in order to be created and configured at startup of the runtime from the build time state. If an object does not have a default constructor you will see an error similar to the following during generation of the augmented artifacts:
```
`have a in order` <- there's a word missing, probably `default constructor`

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+    }
+}
+----
+<1> The serialize method takes the object without a default ctor and creates a `KeyProxy` that contains the information neccessary to recreate the `DSAPublicKey`.
```
s/ctor/constructor/ maybe?
s/neccessary/necessary/

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+}
+----
+<1> The serialize method takes the object without a default ctor and creates a `KeyProxy` that contains the information neccessary to recreate the `DSAPublicKey`.
+<2> The deserialize uses the `KeyProxy` to recreated the `DSAPublicKey` from it's encoded form using the key factory.
```
s/The deserialize/The deserialize method/
s/recreated/recreate/
s/it's/its/

gsmet commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -567,3 +570,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+----
+
+=== Replacing Classes in the Native Image
+The Graal SDK supports subsitutions of classes in the native image. An example of how one could replace the `XmlConfig/XmlData` classes with versions that have no JAXB annotation dependencies is shown in these example classes:
```
s/subsitutions/substitutions/

gsmet reviewed (dismissed):
Nice addition. I added some comments inline.

mkouba commented on docs/src/main/asciidoc/extension-authors-guide.adoc:
```diff
@@ -565,3 +568,483 @@ memory in a runtime class loader. To have these classes written out to disk for
+
+public final class TestProcessor {
+    static DotName TEST_ANNOTATION = DotName.createSimple(TestAnnotation.class.getName());
+    static DotName TEST_ANNOTATION_SCOPE = DotName.createSimple(ApplicationScoped.class.getName());
```
You can replace this with `io.quarkus.arc.processor.BuiltinScope.APPLICATION.getInfo().getDotName()`. Maybe, we should add a convenient method like `BuiltinScope.getName()`. Also the defaultScope is not mandatory - `@Dependent` is used by default.

starksm64 committed:
Updates for the Sample Test Extension section of the extension-authors-guide

Add a code sample section based on the core test-extension

Add a new section that shows various command tasks using the core test-extension code
Updates for the Sample Test Extension section of the extension-authors-guide

Updates for the Sample Test Extension section of the extension-authors-guide

Updates for the Sample Test Extension section of the extension-authors-guide

Updates for the Sample Test Extension section of the extension-authors-guide

Updates for the Sample Test Extension section of the extension-authors-guide

Initial native image test and updates

Additional examples

Reformat

Reformatting

Sketch out object substitution and and replacement

Fix formatting

Rename the native image integration test

Update comment

Update comments

Add sections on deployment scanning and bean container interface

Reformat

Reformat

Don't rely on JDK JAXB so code compiles and runs under both JDK8/11

Updates from review

Fix rebase issues

Updates for the Sample Test Extension section of the extension-authors-guide

Add a code sample section based on the core test-extension

Add a new section that shows various command tasks using the core test-extension code
Updates for the Sample Test Extension section of the extension-authors-guide

Updates for the Sample Test Extension section of the extension-authors-guide

Updates for the Sample Test Extension section of the extension-authors-guide

Updates for the Sample Test Extension section of the extension-authors-guide

Updates for the Sample Test Extension section of the extension-authors-guide

Initial native image test and updates

Additional examples

Reformat

Reformatting

Sketch out object substitution and and replacement

Fix formatting

Rename the native image integration test

Update comment

Update comments

Add sections on deployment scanning and bean container interface

Updates from review

Another rebase attempt

Remove undertow-servlet dependency

Another rebase attempt

Remove the quarkus maven plugin

Update for AdditionalBeanBuildItem changes

starksm64 committed:
Move the native tests out of core to avoid build order problems that hang the integration tests

starksm64 commented:
The CI build failures should not be resolved as the devtools dependency in the core/test-extension have been moved into the integration/test-extension test module

starksm64 committed:
Add javax.activation dependency for Java 11 build

gsmet commented:
@starksm64 it looks like some of my comments were not addressed. Could you take a look, please?

starksm64 committed:
Correct the use of the FeatureBuildItem name

starksm64 committed:
more fixes that I missed comments on

starksm64 commented:
Ok, not sure why I keep missing some of your comments, but I have addressed the majority of them.

Sanne commented:
there's a failure with JDK11:
```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:build (default) on project quarkus-integration-test-elytron-security: Failed to build a runnable JAR: Failed to build a runner jar: Failed to augment application classes: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.elytron.security.deployment.SecurityDeploymentProcessor#registerJCAProviders threw an exception: java.lang.IllegalAccessError: class io.quarkus.elytron.security.deployment.SecurityDeploymentProcessor (in unnamed module @0xad66359) cannot access class sun.security.jca.Providers (in module java.base) because module java.base does not export sun.security.jca to unnamed module @0xad66359

```

starksm64 committed:
Remove unused sun.* imports

starksm64 commented:
Strange, this seemed to pass previously. I have removed that import as it was unused.

dmlloyd commented:
We've just recently become strict about using non-standard classes, in order to prevent build errors of the flavor previously seen when the project is partially built with Java 9 or later and then a native image is produced (unsuccessfully due to referencing Java 9+ symbols and overloads).

starksm64 committed:
Another sun.* package

starksm64 commented:
@Sanne should be good to go

Sanne commented:
thanks @starksm64 . I see you merged master into your feature branch, isn't that confusing? You mind if I rebase it one last time?

starksm64 commented:
No, that is fine. I'm not sure how that happened.

Sanne committed:
Update the groupId for quarkus-bootstrap-maven-plugin in the extensions author guide

gsmet reviewed (approved):
Approved and tested by Sanne.

Sanne commented:
rebased and added a micro update: 49f617404

I'll need @gsmet to merge it as I no longer can push the button. Thanks!
